# Introduction

Selecting specific values of a pandas DataFrame or Series to work on is an implicit step in almost any data operation you'll run, so one of the first things you need to learn in working with data in Python is how to go about selecting the data points relevant to you quickly and effectively.

In [2]:

import pandas as pd
titanic = pd.read_csv("train.csv")


In [4]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [6]:
titanic.shape

(891, 12)

# Native accessors

Native Python objects provide  good ways of indexing data. Pandas carries all of these over, which helps make it easy to start with.



In pandas we can access individual columns by 'indexing'.n use:

In [4]:
titanic.Name

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

If we have a Python dictionary, we can access its values using the indexing (`[]`) operator. We can do the same with columns in a DataFrame:

In [5]:
titanic['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

These are the two ways of selecting a specific Series out of a DataFrame. Neither of them is more or less syntactically valid than the other, but the indexing operator `[]` does have the advantage that it can handle column names with reserved characters in them.

To drill down to a single specific value, we need only use the indexing operator `[]` once more:

In [6]:
titanic['Name'][0]

'Braund, Mr. Owen Harris'

# Indexing in pandas

The indexing operator and attribute selection are nice because they work just like they do in the rest of the Python ecosystem. As a novice, this makes them easy to pick up and use. However, pandas has its own accessor operators, `loc` and `iloc`. For more advanced operations, these are the ones you're supposed to be using.

### Index-based selection

Pandas indexing works in one of two paradigms. The first is **index-based selection**: selecting data based on its numerical position in the data. `iloc` follows this paradigm.

To select the first row of data in a DataFrame, we may use the following:

In [8]:
titanic.iloc[0]

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

Both `loc` and `iloc` are row-first, column-second. 

This means that it's marginally easier to retrieve rows, and marginally harder to get retrieve columns. To get a column with `iloc`, we can do the following:

In [10]:
titanic.iloc[:, 0]

0        1
1        2
2        3
3        4
4        5
      ... 
886    887
887    888
888    889
889    890
890    891
Name: PassengerId, Length: 891, dtype: int64

On its own, the `:` operator, which also comes from native Python, means "everything". When combined with other selectors, however, it can be used to indicate a range of values. For example, to select the `country` column from just the first, second, and third row, we would do:

In [12]:
titanic.iloc[:3, 0]

0    1
1    2
2    3
Name: PassengerId, dtype: int64

Or, to select just the second and third entries, we would do:

In [14]:
titanic.iloc[1:3, 0]

1    2
2    3
Name: PassengerId, dtype: int64

It's also possible to pass a list:

In [16]:
titanic.iloc[[0, 1, 2], 0]

0    1
1    2
2    3
Name: PassengerId, dtype: int64

Finally, it's worth knowing that negative numbers can be used in selection. This will start counting forwards from the _end_ of the values. So for example here are the last five elements of the dataset.

In [18]:
titanic.iloc[-5:]

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

### Label-based selection

The second paradigm for attribute selection is the one followed by the `loc` operator: **label-based selection**. In this paradigm, it's the data index value, not its position, which matters.

For example, to get the first entry, we would now do the following:

In [13]:
titanic.loc[0, 'Name']

'Braund, Mr. Owen Harris'

`iloc` is conceptually simpler than `loc` because it ignores the dataset's indices. When we use `iloc` we treat the dataset like a big matrix (a list of lists), one that we have to index into by position. `loc`, by contrast, uses the information in the indices to do its work. Since your dataset usually has meaningful indices, it's usually easier to do things using `loc` instead. For example, here's one operation that's much easier using `loc`:

In [14]:
titanic.loc[:, ['Name', 'Sex', 'Age']]

Name     Sex   Age
0                              Braund, Mr. Owen Harris    male  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0
2                               Heikkinen, Miss. Laina  female  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0
4                             Allen, Mr. William Henry    male  35.0
..                                                 ...     ...   ...
886                              Montvila, Rev. Juozas    male  27.0
887                       Graham, Miss. Margaret Edith  female  19.0
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN
889                              Behr, Mr. Karl Howell    male  26.0
890                                Dooley, Mr. Patrick    male  32.0

[891 rows x 3 columns]

# Conditional selection

So far we've been indexing various strides of data, using structural properties of the DataFrame itself. To do *interesting* things with the data, however, we often need to ask questions based on conditions. 

For example, suppose that we're interested specifically in better-than-average wines produced in Italy.

We can start by checking how many passengers embarked at "S"

In [20]:
titanic.Embarked == 'S'

0       True
1      False
2       True
3       True
4       True
       ...  
886     True
887     True
888     True
889    False
890    False
Name: Embarked, Length: 891, dtype: bool

This operation produced a Series of `True`/`False` booleans.  This result can then be used inside of `loc` to select the relevant data:

In [22]:
titanic.loc[titanic.Embarked == 'S']

PassengerId  Survived  Pclass  \
0              1         0       3   
2              3         1       3   
3              4         1       1   
4              5         0       3   
6              7         0       1   
..           ...       ...     ...   
883          884         0       2   
884          885         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                         Braund, Mr. Owen Harris    male  22.0      1      0   
2                          Heikkinen, Miss. Laina  female  26.0      0      0   
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                        Allen, Mr. William Henry    male  35.0      0      0   
6                         McCarthy, Mr. Timothy J    male  54.0      0      0   
..                                            ...     ...   ...    ...    ...   
883                 Banfield, Mr. Frederick James    male  28.0      0      0   
884                        Sutehall, Mr. Henry Jr    male  25.0      0      0   
886                         Montvila, Rev. Juozas    male  27.0      0      0   
887                  Graham, Miss. Margaret Edith  female  19.0      0      0   
888      Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1      2   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
6               17463  51.8625   E46        S  
..                ...      ...   ...      ...  
883  C.A./SOTON 34068  10.5000   NaN        S  
884   SOTON/OQ 392076   7.0500   NaN        S  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
888        W./C. 6607  23.4500   NaN        S  

[644 rows x 12 columns]

This DataFrame has 644 rows. The original had 891. About 72% of passengers embarked at S.

We can use the ampersand (`&`) to bring the two questions together.  If we want to know how many passengers who embarked at S also survived.

In [29]:
titanic.loc[(titanic.Embarked == 'S') & (titanic.Survived == 1)]

PassengerId  Survived  Pclass  \
2              3         1       3   
3              4         1       1   
8              9         1       3   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
865          866         1       2   
869          870         1       3   
871          872         1       1   
880          881         1       2   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   
869                    Johnson, Master. Harold Theodor    male   4.0      1   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch            Ticket     Fare Cabin Embarked  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
10       1           PP 9549  16.7000    G6        S  
11       0            113783  26.5500  C103        S  
..     ...               ...      ...   ...      ...  
865      0            236852  13.0000   NaN        S  
869      1            347742  11.1333   NaN        S  
871      1             11751  52.5542   D35        S  
880      1            230433  26.0000   NaN        S  
887      0            112053  30.0000   B42        S  

[217 rows x 12 columns]

Suppose we want to know the passengers who embarked at Q OR has a ticket number = '113803'. For this we use a pipe (`|`):

In [31]:
titanic.loc[(titanic.Embarked == 'Q') | (titanic.Ticket == '113803')]

PassengerId  Survived  Pclass  \
3              4         1       1   
5              6         0       3   
16            17         0       3   
22            23         1       3   
28            29         1       3   
..           ...       ...     ...   
790          791         0       3   
825          826         0       3   
828          829         1       3   
885          886         0       3   
890          891         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
3    Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
5                                Moran, Mr. James    male   NaN      0      0   
16                           Rice, Master. Eugene    male   2.0      4      1   
22                    McGowan, Miss. Anna "Annie"  female  15.0      0      0   
28                  O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0      0   
..                                            ...     ...   ...    ...    ...   
790                      Keane, Mr. Andrew "Andy"    male   NaN      0      0   
825                               Flynn, Mr. John    male   NaN      0      0   
828                  McCormack, Mr. Thomas Joseph    male   NaN      0      0   
885          Rice, Mrs. William (Margaret Norton)  female  39.0      0      5   
890                           Dooley, Mr. Patrick    male  32.0      0      0   

     Ticket     Fare Cabin Embarked  
3    113803  53.1000  C123        S  
5    330877   8.4583   NaN        Q  
16   382652  29.1250   NaN        Q  
22   330923   8.0292   NaN        Q  
28   330959   7.8792   NaN        Q  
..      ...      ...   ...      ...  
790   12460   7.7500   NaN        Q  
825  368323   6.9500   NaN        Q  
828  367228   7.7500   NaN        Q  
885  382652  29.1250   NaN        Q  
890  370376   7.7500   NaN        Q  

[79 rows x 12 columns]